python/lsst/daf/butler/core/datasets/ref.py

```diff
@@ -0,0 +1,278 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ["DatasetRef"]
+
+import hashlib
+from typing import Any, Dict, Mapping, Optional, Tuple
+
+from types import MappingProxyType
+from ..dimensions import DataCoordinate, DimensionGraph, ExpandedDataCoordinate
+from ..run import Run
+from ..configSupport import LookupKey
+from ..utils import immutable
+from .type import DatasetType
+
+
+@immutable
+class DatasetRef:
+    """Reference to a Dataset in a `Registry`.
+
+    A `DatasetRef` may point to a Dataset that currently does not yet exist
+    (e.g., because it is a predicted input for provenance).
+
+    Parameters
+    ----------
+    datasetType : `DatasetType`
+        The `DatasetType` for this Dataset.
+    dataId : `DataCoordinate`
+        A mapping of dimensions that labels the Dataset within a Collection.
+    id : `int`, optional
+        The unique integer identifier assigned when the dataset is created.
+    run : `Run`, optional
+        The run this dataset was associated with when it was created.
+    hash : `bytes`, optional
+        A hash of the dataset type and data ID.  Should only be provided if
+        copying from another `DatasetRef` with the same dataset type and data
+        ID.
+    components : `dict`, optional
+        A dictionary mapping component name to a `DatasetRef` for that
```

Comment:
We don't enforce that the names of these components match the allowed names for the associated StorageClass. I also don't think we check that `run` in the components matches the parent or whether some components have IDs but others don't.

Reply:
:+1: to checking the component names against the storage class.  I don't think the run has to be the same between the parent and components in general (specifically when components are written as separate datasets and then connected into a virtual composite later).  We don't do that in practice right now, and I'm not sure it all works, but I don't want to throw up any road blocks.